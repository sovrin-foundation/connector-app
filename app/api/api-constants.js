// @flow

export const TOKEN_EXPIRED_CODE = 'VES-101'
export const SERVER_ERROR_CODE = 'OCS-000'
export const SERVER_API_CALL_ERROR = 'OCS-001'
export const PENDING_CONNECTION_REQUEST_CODE = 'CRS-103'
export const PUSH_NOTIFICATION_SENT_CODE = 'push-notification-sent'

export const QR_CODE_REQUEST_ID = 'id'
export const QR_CODE_SENDER_DETAIL = 's'
export const QR_CODE_SENDER_DID = 'd'
export const QR_CODE_SENDER_NAME = 'n'
export const QR_CODE_LOGO_URL = 'l'
export const QR_CODE_SENDER_VERIFICATION_KEY = 'v'
export const QR_CODE_SENDER_KEY_DELEGATION = 'dp'
export const QR_CODE_DELEGATION_DID = 'd'
export const QR_CODE_DELEGATION_KEY = 'k'
export const QR_CODE_DELEGATION_SIGNATURE = 's'
export const QR_CODE_SENDER_AGENCY = 'sa'
export const QR_CODE_SENDER_AGENCY_DID = 'd'
export const QR_CODE_SENDER_AGENCY_KEY = 'v'
export const QR_CODE_SENDER_AGENCY_ENDPOINT = 'e'
export const QR_CODE_TARGET_NAME = 't'
export const GENESIS_FILE_NAME = 'pool_transactions_genesis'

export const MESSAGE_TYPE = {
  AUTH: 'auth-req',
  CLAIM_OFFER: 'credOffer',
  CLAIM_REQUEST: 'credReq',
  PROOF_REQUEST: 'proofReq',
  PROOF: 'proof',
  CLAIM: 'cred',
}

export const PAYLOAD_TYPE = {
  GET_MESSAGE: 'GET_MSGS',
  UPDATE_PUSH_COM_METHOD: 'UPDATE_PUSH_COM_METHOD',
  SEND_MESSAGE: 'SEND_MSG',
}

export const STATUS_CODE = {
  ACCEPT: 'MS-104',
}

export const ERROR_ALREADY_EXIST = {
  code: 'OCS',
  message: 'You already have a relationship with ',
}

export const ERROR_ALREADY_EXIST_TITLE = 'Relationship already established'

export const ERROR_INVITATION_RESPONSE_PARSE_CODE = 'RN-100'

export const ERROR_INVITATION_RESPONSE_PARSE =
  'Error parsing error message in invitation response'

export const ERROR_PENDING_INVITATION_RESPONSE_PARSE_CODE = 'RN-101'

export const ERROR_PENDING_INVITATION_RESPONSE_PARSE =
  'Error parsing error message Pending Invitation SMS'

export const ERROR_INVITATION_RESPONSE_FAILED =
  'Failed to establish connection. Please try again later.'
