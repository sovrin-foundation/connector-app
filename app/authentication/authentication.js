// @flow
import React, { PureComponent } from 'react'
import { View, Text, Image } from 'react-native'
import { connect } from 'react-redux'
import { bindActionCreators } from 'redux'
import { connectionRoute, homeRoute } from '../common'
import { getConnection } from '../store/connections-store'
import type { Store } from '../store/type-store'
import { Container, Request } from '../components'
import ConnectionSuccessModal from './connection-success-modal'
import {
  AUTHENTICATION_STATUS,
  resetAuthenticationStatus,
  sendUserAuthenticationResponse,
} from './authentication-store'
import type { Connection } from '../store/type-connection-store'

// TODO:KS Fix type for this screen, we are not using this screen
// as of now, so we will fix this later
class Authentication extends PureComponent<any, any> {
  state = {
    isModalVisible: false,
  }

  _showConnectionSuccessModal = (isModalVisible, route) => {
    this.setState({ isModalVisible })
    if (route) {
      this.props.navigation.navigate(route)
    }
  }

  componentWillReceiveProps(nextProps) {
    if (nextProps.authentication.status != this.props.authentication.status) {
      if (
        !nextProps.authentication.isFetching &&
        nextProps.authentication.status === AUTHENTICATION_STATUS.ACCEPTED &&
        !nextProps.authentication.error
      ) {
        this.props.navigation.navigate(connectionRoute)
        this.props.resetAuthenticationStatus()
      } else if (
        !nextProps.authentication.isFetching &&
        (nextProps.authentication.status === AUTHENTICATION_STATUS.REJECTED ||
          nextProps.authentication.error)
      ) {
        this.props.navigation.navigate(homeRoute)
        this.props.resetAuthenticationStatus()
      }
    }
  }

  onUserResponse = (newStatus: string) => {
    let {
      type: authenticationType,
      data: { remoteConnectionId },
    } = this.props.authentication

    const { connections: { data: connectionsData } } = this.props
    // $FlowFixMe
    const connection: Connection[] = getConnection(
      remoteConnectionId,
      connectionsData
    )
    let { identifier } = connection[0]

    const challenge = JSON.stringify({
      newStatus,
    })

    this.props.sendUserAuthenticationResponse(
      {
        newStatus,
        identifier,
        remoteConnectionId,
        dataBody: {
          challenge,
        },
      },
      this.props.config,
      authenticationType
    )
  }

  render() {
    const { data } = this.props.authentication
    let offerMsgTitle = '',
      offerMsgText = '',
      logoUrl = null,
      name = ''

    if (data) {
      offerMsgTitle = data.offerMsgTitle
      offerMsgText = data.offerMsgText
      logoUrl = data.logoUrl
    }

    return (
      <Container>
        <Request
          title={offerMsgTitle}
          message={offerMsgText}
          senderLogoUrl={logoUrl}
          onAction={this.onUserResponse}
          navigation={this.props.navigation}
          showErrorAlerts={this.props.config.showErrorAlerts}
          testID="authentication-request"
          senderName={name}
        />
        <ConnectionSuccessModal
          isModalVisible={this.state.isModalVisible}
          showConnectionSuccessModal={this._showConnectionSuccessModal}
          name={name}
          logoUrl={logoUrl}
        />
      </Container>
    )
  }
}

const mapStateToProps = ({ authentication, connections, config }: Store) => ({
  authentication,
  connections,
  config,
})

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      resetAuthenticationStatus,
      sendUserAuthenticationResponse,
    },
    dispatch
  )

export default connect(mapStateToProps, mapDispatchToProps)(Authentication)
