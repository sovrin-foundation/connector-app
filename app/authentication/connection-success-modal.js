// @flow
import React, { PureComponent } from 'react'
import { View, Image, StyleSheet } from 'react-native'

import { CustomModal, AvatarsPair, CustomText } from '../components'
import { color, OFFSET_1X, OFFSET_2X } from '../common/styles'
import { connectionRoute } from '../common'
import { getConnectionLogo } from '../store/connections-store'

// TODO:KS Move these props to type-* file
// we will move this file as well to components directory
// and create tests, type and component in separate folder
// Trying to avoid as many conflicts as possible
export type ConnectionSuccessModalProps = {
  name: string,
  logoUrl: ?string,
  showConnectionSuccessModal: (boolean, string) => void,
  isModalVisible: boolean,
}

export default class ConnectionSuccessModal extends PureComponent<
  ConnectionSuccessModalProps,
  void
> {
  render() {
    const {
      name = 'Sovrin',
      logoUrl,
      showConnectionSuccessModal,
      isModalVisible,
    } = this.props
    const connectionAvatar = getConnectionLogo(logoUrl)

    return (
      <CustomModal
        onPress={() => showConnectionSuccessModal(false, connectionRoute)}
        buttonText="Continue"
        testID={'invitation'}
        isVisible={isModalVisible}
      >
        <AvatarsPair
          middleImage={require('../images/checkMark.png')}
          avatarRight={connectionAvatar}
          testID={'invitation'}
        />
        <CustomText
          bg="tertiary"
          h5
          tertiary
          demiBold
          center
          style={[styles.message]}
          transparentBg
          testID={`invitation-message`}
        >
          {`You are now connected to ${name}!`}
        </CustomText>
      </CustomModal>
    )
  }
}

const styles = StyleSheet.create({
  message: {
    paddingTop: OFFSET_1X,
    marginBottom: OFFSET_1X,
    marginHorizontal: OFFSET_1X / 2,
  },
})
