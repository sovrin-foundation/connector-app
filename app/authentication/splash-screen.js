// @flow
import React, { PureComponent } from 'react'
import { connect } from 'react-redux'
import { View } from 'react-native'
import { bindActionCreators } from 'redux'
import SplashScreen from 'react-native-splash-screen'
import {
  homeRoute,
  splashScreenRoute,
  expiredTokenRoute,
  lockSelectionRoute,
  lockEnterPinRoute,
  lockEnterFingerprintRoute,
  invitationRoute,
  waitForInvitationRoute,
  eulaRoute,
  restoreRoute,
} from '../common/route-constants'
import { Container, Loader } from '../components'
import {
  TOKEN_EXPIRED_CODE,
  PENDING_CONNECTION_REQUEST_CODE,
} from '../api/api-constants'
import { addPendingRedirection } from '../lock/lock-store'
import {
  getSmsPendingInvitation,
  safeToDownloadSmsInvitation,
} from '../sms-pending-invitation/sms-pending-invitation-store'
import type { SplashScreenProps } from './type-splash-screen'
import type { Store } from '../store/type-store'
import { SMSPendingInvitationStatus } from '../sms-pending-invitation/type-sms-pending-invitation'
import type { SMSPendingInvitations } from '../sms-pending-invitation/type-sms-pending-invitation'
import { deepLinkProcessed } from '../deep-link/deep-link-store'
import { DEEP_LINK_STATUS } from '../deep-link/type-deep-link'

const isReceived = ({ payload, status }) =>
  payload &&
  payload.senderDetail.DID &&
  status === SMSPendingInvitationStatus.RECEIVED

export class SplashScreenView extends PureComponent<SplashScreenProps, void> {
  ifDeepLinkFoundGoToWaitForInvitationScreenNFetchInvitation = (
    prevProps: SplashScreenProps
  ) => {
    const nextDeepLinkTokens = this.props.deepLink.tokens
    if (
      this.props.deepLink.isLoading === false &&
      Object.keys(nextDeepLinkTokens).length >
        Object.keys(prevProps.deepLink.tokens).length
    ) {
      Object.keys(nextDeepLinkTokens).map(
        smsToken =>
          nextDeepLinkTokens[smsToken].status !== DEEP_LINK_STATUS.PROCESSED &&
          this.props.getSmsPendingInvitation(smsToken)
      )
      if (this.props.lock.isAppLocked === false) {
        this.props.navigation.navigate(waitForInvitationRoute)
      }
    }
  }

  getUnHandledSmsPendingInvitations = () => {
    const smsPendingInvitations = Object.keys(
      this.props.smsPendingInvitation
    ).map(invitationToken => {
      return {
        ...this.props.smsPendingInvitation[invitationToken],
        invitationToken,
      }
    })
    const unHandledSmsPendingInvitations: SMSPendingInvitations = smsPendingInvitations.filter(
      ({ invitationToken }) =>
        invitationToken &&
        this.props.deepLink.tokens[invitationToken].status !==
          DEEP_LINK_STATUS.PROCESSED
    )
    return unHandledSmsPendingInvitations
  }

  handleExpiredTokens = (
    unHandledSmsPendingInvitations: SMSPendingInvitations
  ) => {
    const isAnyOneOfSmsPendingInvitationHasError: boolean = unHandledSmsPendingInvitations.some(
      ({ error }) => error
    )
    const isAnyOneOfSmsPendingInvitationWasExpired: boolean = unHandledSmsPendingInvitations.some(
      ({ error }) => error && error.code && error.code === TOKEN_EXPIRED_CODE
    )
    if (isAnyOneOfSmsPendingInvitationWasExpired) {
      // TODO: This conditions pops up too many times, we should create a common function for this.
      // TODO: It should be something like this this.redirect(nextProps.lock, routeName)
      if (this.props.lock.isAppLocked === false) {
        this.props.navigation.navigate(expiredTokenRoute)
      } else {
        this.props.addPendingRedirection([{ routeName: expiredTokenRoute }])
      }
    } else if (isAnyOneOfSmsPendingInvitationHasError) {
      // * This condition is needed to avoid un wanted redirection to home screen if unHandledSmsPendingInvitations are empty
      // * or if we are in middle of other invitation fetching process
      if (this.props.lock.isAppLocked === false) {
        this.props.navigation.navigate(homeRoute)
      } else {
        this.props.addPendingRedirection([{ routeName: homeRoute }])
      }
    }
    unHandledSmsPendingInvitations.map(({ error, invitationToken }) => {
      error ? this.props.deepLinkProcessed(invitationToken) : null
    })
  }

  handleSmsPendingInvitations = (prevProps: SplashScreenProps) => {
    const nextDeepLinkTokens = this.props.deepLink.tokens
    if (
      JSON.stringify(prevProps.smsPendingInvitation) !==
      JSON.stringify(this.props.smsPendingInvitation)
    ) {
      const unHandledSmsPendingInvitations = this.getUnHandledSmsPendingInvitations()
      this.handleExpiredTokens(unHandledSmsPendingInvitations)
      const unseenSmsPendingInvitations = unHandledSmsPendingInvitations.filter(
        isReceived
      )

      const pendingRedirectionList = unseenSmsPendingInvitations.map(
        ({ payload, invitationToken }) => {
          if (payload) {
            const senderDID = payload.senderDetail.DID
            this.props.deepLinkProcessed(invitationToken)
            if (this.props.lock.isAppLocked === false) {
              this.props.navigation.push(invitationRoute, {
                senderDID,
                token: invitationToken,
              })
            }
            return {
              routeName: invitationRoute,
              params: { senderDID, token: invitationToken },
            }
          }
        }
      )

      pendingRedirectionList.length !== 0 &&
        this.props.lock.isAppLocked === true &&
        this.props.addPendingRedirection(pendingRedirectionList)

      // * all error token links should be processed
    }
  }

  componentDidUpdate(prevProps: SplashScreenProps) {
    if (this.props.isInitialized !== prevProps.isInitialized) {
      // hydrated is changed, and if it is changed to true,
      // that means this is the only time we would get inside this if condition
      this.initializationCheck()
    }

    // check if deepLink is changed, then that means we either got token
    // or we got error or nothing happened with deep link
    const nextDeepLinkTokens = this.props.deepLink.tokens
    if (
      prevProps.deepLink.isLoading !== this.props.deepLink.isLoading &&
      this.props.deepLink.isLoading === false &&
      Object.keys(nextDeepLinkTokens).length === 0
    ) {
      if (this.props.lock.isAppLocked === false) {
        // we did not get any token and deepLink data loading is done
        SplashScreen.hide()
        this.props.navigation.navigate(homeRoute)
      } else {
        this.props.addPendingRedirection([{ routeName: homeRoute }])
      }
    }

    this.ifDeepLinkFoundGoToWaitForInvitationScreenNFetchInvitation(prevProps)
    this.handleSmsPendingInvitations(prevProps)
  }

  componentDidMount() {
    // if the SplashScreen component has been mounted for 2 seconds, hide it to render the Loading component
    global.setTimeout(() => {
      SplashScreen.hide()
    }, 2000)
    // It might be the case the hydration finishes
    // even before component is mounted,
    // so we need to check for pin code here as well

    this.initializationCheck()
  }

  initializationCheck() {
    if (this.props.isInitialized) {
      SplashScreen.hide()
      // now we can safely check value of isAlreadyInstalled
      if (
        !this.props.lock.isLockEnabled ||
        this.props.lock.isLockEnabled === 'false'
      ) {
        // user is opening the app for first time after installing
        if (!this.props.eula.isEulaAccept) {
          this.props.navigation.navigate(eulaRoute)
        } else {
          this.props.navigation.navigate(restoreRoute)
        }
      } else {
        // not the first time user is opening app
        if (this.props.lock.isTouchIdEnabled) {
          this.props.navigation.navigate(lockEnterFingerprintRoute)
        } else {
          this.props.navigation.navigate(lockEnterPinRoute)
        }
      }
    }
  }

  // conditional render is to eliminate InteractionManager.allowInteractions issues
  render() {
    return this.props.isInitialized ? null : (
      <Container center>
        <Loader />
      </Container>
    )
  }
}

const mapStateToProps = ({
  config,
  deepLink,
  lock,
  smsPendingInvitation,
  eula,
}: Store) => ({
  isInitialized: config.isInitialized,
  // DeepLink should be it's own component that will handle only deep link logic
  // in that way, we will be able to restrict re-render and re-run of code
  deepLink,
  // only need 3 props
  lock,
  smsPendingInvitation,
  // only need isEulaAccept
  eula,
})

const mapDispatchToProps = dispatch =>
  bindActionCreators(
    {
      getSmsPendingInvitation,
      addPendingRedirection,
      safeToDownloadSmsInvitation,
      deepLinkProcessed,
    },
    dispatch
  )

export default connect(mapStateToProps, mapDispatchToProps)(SplashScreenView)
