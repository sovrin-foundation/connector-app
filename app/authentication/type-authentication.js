// @flow
import type { ConfigStore } from '../store/type-config-store'

export type AuthenticationStatus = {
  ACCEPTED: string,
  REJECTED: string,
  NONE: string,
}

export type AuthenticationType = {
  NONE: string,
  AUTHENTICATION_REQUEST: string,
}

export type AuthenticationError = {
  message: string,
  authenticationType: string,
}

type Invitee = {
  +image: string,
}

type Inviter = {
  +image: string,
}

type DataBody = {
  challenge: string,
  signature?: string,
}

export type AuthenticationSuccessData = {
  newStatus: string,
  identifier: string,
  dataBody: DataBody,
  remoteConnectionId: string,
}

export type AuthenticationStore = {
  +inviter: Inviter,
  +invitee: Invitee,
  +data: ?AuthenticationPayload,
  +error: ?AuthenticationError,
  +type: string,
  +status: string,
  +isFetching: boolean,
  +isPristine: boolean,
}

export type AuthenticationPayload =
  | {
      offerMsgTitle: string,
      offerMsgText: string,
      logoUrl: ?string,
      name?: string,
      statusMsg?: string,
      remoteConnectionId: string,
    }
  | AuthenticationSuccessData

export type NewConnection = {
  identifier: string,
  remoteConnectionId: string,
  seed: string,
  remoteDID: string,
}

export type AuthenticationAction =
  | SendUserAuthenticationResponse
  | SendUserAuthenticationResponseSuccess
  | SendUserAuthenticationResponseFailure
  | AuthenticationRequestReceived
  | ResetAuthenticationStatus

const AUTHENTICATION_REQUEST_RECEIVED = 'AUTHENTICATION_REQUEST_RECEIVED'
const SEND_USER_AUTHENTICATION_RESPONSE = 'SEND_USER_AUTHENTICATION_RESPONSE'
const SEND_USER_AUTHENTICATION_RESPONSE_SUCCESS =
  'SEND_USER_AUTHENTICATION_RESPONSE_SUCCESS'
const SEND_USER_AUTHENTICATION_RESPONSE_FAILURE =
  'SEND_USER_AUTHENTICATION_RESPONSE_FAILURE'
const RESET_AUTHENTICATION_STATUS = 'RESET_AUTHENTICATION_STATUS'

export type SendUserAuthenticationResponse = {
  type: typeof SEND_USER_AUTHENTICATION_RESPONSE,
  data: AuthenticationSuccessData,
  config: ConfigStore,
  authenticationType: string,
  token?: string,
  newConnection?: NewConnection,
}

export type SendUserAuthenticationResponseSuccess = {
  type: typeof SEND_USER_AUTHENTICATION_RESPONSE_SUCCESS,
  data: AuthenticationSuccessData,
}

export type SendUserAuthenticationResponseFailure = {
  type: typeof SEND_USER_AUTHENTICATION_RESPONSE_FAILURE,
  error: AuthenticationError,
}

export type AuthenticationRequestReceived = {
  type: typeof AUTHENTICATION_REQUEST_RECEIVED,
  data: AuthenticationPayload,
}

export type ResetAuthenticationStatus = {
  type: typeof RESET_AUTHENTICATION_STATUS,
}
