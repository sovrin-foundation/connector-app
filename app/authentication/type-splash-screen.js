// @flow

import type { ReactNavigation } from '../common/type-common'
import type { ConfigStore } from '../store/type-config-store'
import type { DeepLinkStore } from '../deep-link/type-deep-link'
import type { SMSPendingInvitationStore } from '../sms-pending-invitation/type-sms-pending-invitation'
import type { LockStore, PendingRedirection } from '../lock/type-lock'
import type { EulaStore } from '../eula/type-eula'

export type SplashScreenProps = {
  eula: EulaStore,
  isInitialized: boolean,
  deepLink: DeepLinkStore,
  smsPendingInvitation: SMSPendingInvitationStore,
  lock: LockStore,
  getSmsPendingInvitation: (token: string) => void,
  addPendingRedirection: (
    pendingRedirection: Array<?PendingRedirection>
  ) => void,
  safeToDownloadSmsInvitation: () => void,
  deepLinkProcessed: (data: string) => void,
} & ReactNavigation
