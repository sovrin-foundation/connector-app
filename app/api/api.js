// @flow
import { api, options } from './api-utils'
import type {
  GetInvitationLinkApiData,
  MessageApiData,
  DownloadInvitationApiData,
  SendAuthenticationResponseApiData,
  EnvironmentDetailUrlDownloaded,
} from './type-api'
import { PAYLOAD_TYPE, MESSAGE_TYPE, STATUS_CODE } from './api-constants'

export const getInvitationLink = ({
  agencyUrl,
  smsToken,
}: GetInvitationLinkApiData) =>
  api(`${agencyUrl}/agency/url-mapper/${smsToken}`, options('GET'))

export const invitationDetailsRequest = ({ url }: DownloadInvitationApiData) =>
  api(url, options('GET'))

export const getAdditionalData = ({ agencyUrl, dataBody }: MessageApiData) =>
  api(`${agencyUrl}/agency/route`, options('POST', dataBody))

export const sendAuthenticationRequest = ({
  data: { identifier, dataBody },
  config: { agencyUrl },
}: SendAuthenticationResponseApiData) =>
  api(`${agencyUrl}/agent/${identifier}/auth`, options('PUT', dataBody), true)

export const downloadEnvironmentDetails = (
  url: string
): Promise<EnvironmentDetailUrlDownloaded> => api(url, options())
