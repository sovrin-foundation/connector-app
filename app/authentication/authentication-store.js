// @flow
import { put, takeLatest, call, all } from 'redux-saga/effects'
import { delay } from 'redux-saga'
import { sendAuthenticationRequest } from '../api/api'
import type { ConfigStore } from '../store/type-config-store'
import type {
  AuthenticationStatus,
  AuthenticationType,
  AuthenticationStore,
  AuthenticationPayload,
  AuthenticationError,
  AuthenticationAction,
  SendUserAuthenticationResponse,
  AuthenticationSuccessData,
  SendUserAuthenticationResponseSuccess,
  SendUserAuthenticationResponseFailure,
  ResetAuthenticationStatus,
  AuthenticationRequestReceived,
} from './type-authentication'
import { RESET } from '../common/type-common'

export const AUTHENTICATION_STATUS: AuthenticationStatus = {
  ACCEPTED: 'accepted',
  REJECTED: 'rejected',
  NONE: 'NONE',
}

export const AUTHENTICATION_TYPE: AuthenticationType = {
  NONE: 'NONE',
  AUTHENTICATION_REQUEST: 'AUTHENTICATION_REQUEST',
}

const initialState: AuthenticationStore = {
  inviter: { image: '' },
  invitee: { image: './images/inviter.jpeg' },
  data: null,
  error: null,
  type: AUTHENTICATION_TYPE.NONE,
  status: AUTHENTICATION_STATUS.NONE,
  isFetching: false,
  isPristine: true,
}

const AUTHENTICATION_REQUEST_RECEIVED = 'AUTHENTICATION_REQUEST_RECEIVED'
const SEND_USER_AUTHENTICATION_RESPONSE = 'SEND_USER_AUTHENTICATION_RESPONSE'
const SEND_USER_AUTHENTICATION_RESPONSE_SUCCESS =
  'SEND_USER_AUTHENTICATION_RESPONSE_SUCCESS'
const SEND_USER_AUTHENTICATION_RESPONSE_FAILURE =
  'SEND_USER_AUTHENTICATION_RESPONSE_FAILURE'
const RESET_AUTHENTICATION_STATUS = 'RESET_AUTHENTICATION_STATUS'

export const sendUserAuthenticationResponse = (
  data: AuthenticationSuccessData,
  config: ConfigStore,
  authenticationType: string
): SendUserAuthenticationResponse => ({
  type: SEND_USER_AUTHENTICATION_RESPONSE,
  data,
  config,
  authenticationType,
})

export const sendUserAuthenticationResponseSuccess = (
  data: AuthenticationSuccessData
): SendUserAuthenticationResponseSuccess => ({
  type: SEND_USER_AUTHENTICATION_RESPONSE_SUCCESS,
  data,
})

export const sendUserAuthenticationResponseFailure = (
  error: AuthenticationError
): SendUserAuthenticationResponseFailure => ({
  type: SEND_USER_AUTHENTICATION_RESPONSE_FAILURE,
  error,
})

export const authenticationRequestReceived = (
  data: AuthenticationPayload
): AuthenticationRequestReceived => ({
  type: AUTHENTICATION_REQUEST_RECEIVED,
  data,
})

export const resetAuthenticationStatus = (): ResetAuthenticationStatus => ({
  type: RESET_AUTHENTICATION_STATUS,
})

export function* watchLoadAuthenticationDetailsRequest(): any {
  yield takeLatest(
    SEND_USER_AUTHENTICATION_RESPONSE,
    handleUserAuthenticationResponse
  )
}

function* handleUserAuthenticationResponse(
  action: SendUserAuthenticationResponse
): Generator<*, *, *> {
  const { authenticationType } = action
  try {
    let authenticationActionResponse = null
    if (authenticationType == AUTHENTICATION_TYPE.AUTHENTICATION_REQUEST) {
      const { remoteConnectionId, dataBody: { challenge } } = action.data
      // TODO: Fix authentication flow
      const signature = ''
      // add signature
      action.data.dataBody.signature = signature

      // TODO: Fix this while fixing authentication flow
      // authenticationActionResponse = yield call(
      //   sendAuthenticationRequest,
      //   action
      // )
      // yield put(sendUserAuthenticationResponseSuccess(action.data))
    }
  } catch (e) {
    yield put(
      sendUserAuthenticationResponseFailure({
        message: e.message,
        authenticationType: action.authenticationType,
      })
    )
  }
}

export function* watchAuthentication(): any {
  yield all([watchLoadAuthenticationDetailsRequest()])
}

export default function authentication(
  state: AuthenticationStore = initialState,
  action: AuthenticationAction
): AuthenticationStore {
  switch (action.type) {
    case SEND_USER_AUTHENTICATION_RESPONSE:
      return {
        ...state,
        isFetching: true,
        isPristine: false,
      }
    case SEND_USER_AUTHENTICATION_RESPONSE_SUCCESS:
      return {
        ...state,
        isFetching: false,
        status: action.data.newStatus,
      }
    case SEND_USER_AUTHENTICATION_RESPONSE_FAILURE:
      return {
        ...state,
        isFetching: false,
        error: action.error,
      }
    case RESET_AUTHENTICATION_STATUS:
      return {
        ...state,
        type: AUTHENTICATION_TYPE.NONE,
        status: AUTHENTICATION_STATUS.NONE,
        error: null,
        data: null,
      }
    case AUTHENTICATION_REQUEST_RECEIVED:
      return {
        ...state,
        type: AUTHENTICATION_TYPE.AUTHENTICATION_REQUEST,
        status: AUTHENTICATION_STATUS.NONE,
        data: action.data,
        error: null,
      }
    case RESET:
      return initialState
    default:
      return state
  }
}
