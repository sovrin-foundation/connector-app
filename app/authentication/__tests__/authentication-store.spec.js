// @flow
import authentication, {
  sendUserAuthenticationResponseSuccess,
  sendUserAuthenticationResponseFailure,
  authenticationRequestReceived,
  resetAuthenticationStatus,
  sendUserAuthenticationResponse,
  AUTHENTICATION_TYPE,
  AUTHENTICATION_STATUS,
} from '../authentication-store'
import type {
  AuthenticationStatus,
  AuthenticationType,
  AuthenticationStore,
  AuthenticationPayload,
  NewConnection,
  AuthenticationError,
  ResetAuthenticationStatus,
  AuthenticationAction,
  SendUserAuthenticationResponse,
  AuthenticationSuccessData,
  SendUserAuthenticationResponseSuccess,
  SendUserAuthenticationResponseFailure,
  AuthenticationRequestReceived,
} from '../type-authentication'
import { PUSH_NOTIFICATION_SENT_CODE } from '../../api/api-constants'
import { RESET } from '../../common/type-common'
import { getStore } from '../../../__mocks__/static-data'

describe('Authentication request tests', () => {
  const initialAction: AuthenticationAction = {
    type: 'INITIAL_TEST_ACTION',
  }

  let initialState
  beforeEach(() => {
    initialState = authentication(undefined, initialAction)
  })

  it('should correctly match initial state', () => {
    expect(initialState).toMatchSnapshot()
  })

  it('send user authentication response success action should update store', () => {
    const successData = {
      newStatus: AUTHENTICATION_STATUS.ACCEPTED,
      dataBody: { challenge: 'challenge', signature: 'signature' },
      identifier: 'identifier',
      remoteConnectionId: 'remoteConnectionId',
    }
    const successAction: SendUserAuthenticationResponseSuccess = sendUserAuthenticationResponseSuccess(
      successData
    )
    expect(authentication(initialState, successAction)).toMatchSnapshot()
  })

  it('send user authentication response failure action should update store', () => {
    const error = {
      message: '',
      authenticationType: AUTHENTICATION_TYPE.AUTHENTICATION_REQUEST,
    }
    const errorAction: SendUserAuthenticationResponseFailure = sendUserAuthenticationResponseFailure(
      error
    )
    expect(authentication(initialState, errorAction)).toMatchSnapshot()
  })

  it('authentication status reset action should update store', () => {
    const resetAction: ResetAuthenticationStatus = resetAuthenticationStatus()
    expect(authentication(initialState, resetAction)).toMatchSnapshot()
  })

  it('send user authentication response received action should update store', () => {
    const receivedData = {
      offerMsgTitle: 'Hi There',
      offerMsgText: 'Suncoast Credit Union (sandbox) wants to connect with you',
      statusCode: PUSH_NOTIFICATION_SENT_CODE,
      logoUrl: null,
      remoteConnectionId: 'B4Y9fhpeHdGHBKKtSgAYrB',
    }
    const receivedAction: AuthenticationRequestReceived = authenticationRequestReceived(
      receivedData
    )
    expect(authentication(initialState, receivedAction)).toMatchSnapshot()
  })

  it('send user authentication response', () => {
    const newStatus = 'some status'
    const challenge = 'some challenge'
    const signature = 'some signature'
    const identifier = 'some identifier'
    const config = getStore().getState().config

    const responseAction: SendUserAuthenticationResponse = sendUserAuthenticationResponse(
      {
        newStatus,
        identifier,
        remoteConnectionId: 'remoteConnectionId',
        dataBody: {
          challenge,
          signature,
        },
      },
      config,
      AUTHENTICATION_TYPE.AUTHENTICATION_REQUEST
    )
    expect(authentication(initialState, responseAction)).toMatchSnapshot()
  })

  it('should reset authentication store when reset action is raised', () => {
    const successData = {
      newStatus: AUTHENTICATION_STATUS.ACCEPTED,
      dataBody: { challenge: 'challenge', signature: 'signature' },
      identifier: 'identifier',
      remoteConnectionId: 'remoteConnectionId',
    }
    const successState = authentication(
      initialState,
      sendUserAuthenticationResponseSuccess(successData)
    )
    expect(authentication(successState, { type: RESET })).toMatchSnapshot()
  })
})
