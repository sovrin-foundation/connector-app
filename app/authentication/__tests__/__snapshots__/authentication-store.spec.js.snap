// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authentication request tests authentication status reset action should update store 1`] = `
Object {
  "data": null,
  "error": null,
  "invitee": Object {
    "image": "./images/inviter.jpeg",
  },
  "inviter": Object {
    "image": "",
  },
  "isFetching": false,
  "isPristine": true,
  "status": "NONE",
  "type": "NONE",
}
`;

exports[`Authentication request tests send user authentication response 1`] = `
Object {
  "data": null,
  "error": null,
  "invitee": Object {
    "image": "./images/inviter.jpeg",
  },
  "inviter": Object {
    "image": "",
  },
  "isFetching": true,
  "isPristine": false,
  "status": "NONE",
  "type": "NONE",
}
`;

exports[`Authentication request tests send user authentication response failure action should update store 1`] = `
Object {
  "data": null,
  "error": Object {
    "authenticationType": "AUTHENTICATION_REQUEST",
    "message": "",
  },
  "invitee": Object {
    "image": "./images/inviter.jpeg",
  },
  "inviter": Object {
    "image": "",
  },
  "isFetching": false,
  "isPristine": true,
  "status": "NONE",
  "type": "NONE",
}
`;

exports[`Authentication request tests send user authentication response received action should update store 1`] = `
Object {
  "data": Object {
    "logoUrl": null,
    "offerMsgText": "Suncoast Credit Union (sandbox) wants to connect with you",
    "offerMsgTitle": "Hi There",
    "remoteConnectionId": "B4Y9fhpeHdGHBKKtSgAYrB",
    "statusCode": "push-notification-sent",
  },
  "error": null,
  "invitee": Object {
    "image": "./images/inviter.jpeg",
  },
  "inviter": Object {
    "image": "",
  },
  "isFetching": false,
  "isPristine": true,
  "status": "NONE",
  "type": "AUTHENTICATION_REQUEST",
}
`;

exports[`Authentication request tests send user authentication response success action should update store 1`] = `
Object {
  "data": null,
  "error": null,
  "invitee": Object {
    "image": "./images/inviter.jpeg",
  },
  "inviter": Object {
    "image": "",
  },
  "isFetching": false,
  "isPristine": true,
  "status": "accepted",
  "type": "NONE",
}
`;

exports[`Authentication request tests should correctly match initial state 1`] = `
Object {
  "data": null,
  "error": null,
  "invitee": Object {
    "image": "./images/inviter.jpeg",
  },
  "inviter": Object {
    "image": "",
  },
  "isFetching": false,
  "isPristine": true,
  "status": "NONE",
  "type": "NONE",
}
`;

exports[`Authentication request tests should reset authentication store when reset action is raised 1`] = `
Object {
  "data": null,
  "error": null,
  "invitee": Object {
    "image": "./images/inviter.jpeg",
  },
  "inviter": Object {
    "image": "",
  },
  "isFetching": false,
  "isPristine": true,
  "status": "NONE",
  "type": "NONE",
}
`;
